openverse_catalog/dags/providers/provider_api_scripts/inaturalist.py

```diff
@@ -0,0 +1,164 @@
+        if prev_query_params is None:
+            return {"offset_num": 0}
+        else:
+            next_offset = prev_query_params["offset_num"] + self.batch_limit
+            return {"offset_num": next_offset}
+
+    def get_response_json(self, query_params: Dict):
+        """
+        Call the SQL to pull json from Postgres, where the raw data has been loaded.
+        """
+        sql_string = self.sql_template.format(
+            batch_limit=self.batch_limit, offset_num=query_params["offset_num"]
+        )
+        sql_result = self.pg.get_records(sql_string)
+        # Postgres returns a a list of tuples, even if it's one tuple with one item.
+        return sql_result[0][0]
+
+    def get_batch_data(self, response_json):
+        if response_json:
+            return response_json
+        return None
+
+    def get_record_data(self, data):
+        if data.get("foreign_identifier") is None:
+            return None
+        license_url = data.get("license_url")
+        license_info = get_license_info(license_url=license_url)
+        if license_info == LicenseInfo(None, None, None, None):
+            return None
+        record_data = {k: data[k] for k in data.keys() if k != "license_url"}
+        record_data["license_info"] = license_info
+        return record_data
+
+    def get_media_type(self, record):
+        # This provider only supports Images via S3, though they have some audio files
+        # on site and in the API.
+        return "image"
+
+    def endpoint(self):
+        raise NotImplementedError("Normalized TSV files from AWS S3 means no endpoint.")
+
+    @staticmethod
+    def compare_update_dates(
+        last_success: pendulum.DateTime | None, s3_keys: list, aws_conn_id=AWS_CONN_ID
+    ):
+        # if it was never run, assume the data is new
+        if last_success is None:
+            return
+        s3 = S3Hook(aws_conn_id=aws_conn_id)
+        s3_client = s3.get_client_type()
+        for key in s3_keys:
+            # this will error out if the files don't exist, and bubble up as an
+            # informative failure
+            last_modified = s3_client.head_object(
+                Bucket="inaturalist-open-data", Key=key
+            )["LastModified"]
+            # if any file has been updated, let's pull them all
+            if last_success < last_modified:
+                return
+        # If no files have been updated, skip the DAG
+        raise AirflowSkipException("Nothing new to ingest")
+
+    @staticmethod
+    def create_preingestion_tasks():
+
+        with TaskGroup(group_id="preingestion_tasks") as preingestion_tasks:
+
+            check_for_file_updates = PythonOperator(
+                task_id="check_for_file_updates",
+                python_callable=INaturalistDataIngester.compare_update_dates,
+                op_kwargs={
+                    # With the templated values ({{ x }}) airflow will fill it in
```

Comment:
Is this describing a special case or just a helpful note? This was my understanding of how templates worked generally.

Reply:
Right, yes. I was more trying to be clear about where `prev_start_date_success` comes from, but I see how that's maybe obvious and/or not adding enough additional info to be helpful. How about I drop it?

Reply:
Up to you :+1: